osu.Game/Rulesets/Difficulty/Utils/ReverseQueue.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace osu.Game.Rulesets.Difficulty.Utils
+{
+    /// <summary>
+    /// An indexed queue where items are indexed beginning from the end instead of the start.
+    /// </summary>
+    public class ReverseQueue<T> : IEnumerable<T>
+    {
+        /// <summary>
+        /// The number of elements in the <see cref="ReverseQueue{T}"/>.
+        /// </summary>
+        public int Count { get; private set; }
+
+        private T[] items;
+        private int capacity;
+        private int start;
+
+        public ReverseQueue(int initialCapacity)
+        {
+            if (initialCapacity < 0)
+                throw new ArgumentOutOfRangeException(nameof(initialCapacity));
+
+            items = new T[initialCapacity];
+            capacity = initialCapacity;
+            start = 0;
+            Count = 0;
+        }
+
+        /// <summary>
+        /// Retrieves the item at an index in the <see cref="ReverseQueue{T}"/>.
+        /// </summary>
+        /// <param name="index">The index of the item to retrieve. The most recently enqueued item is at index 0.</param>
+        public T this[int index]
+        {
+            get
+            {
+                if (index > Count - 1)
+                    throw new ArgumentOutOfRangeException(nameof(index));
+
+                int reverseIndex = Count - 1 - index;
+                return items[(start + reverseIndex) % capacity];
+            }
+        }
+
+        /// <summary>
+        /// Enqueues an item to this <see cref="ReverseQueue{T}"/>.
+        /// </summary>
+        /// <param name="item">The item to enqueue.</param>
+        public void Enqueue(T item)
+        {
+            if (Count == capacity)
+            {
+                // Double the buffer size
+                var buffer = new T[capacity * 2];
```

Comment:
This fails if `capacity == 0` (which it can be, in particular initially).

Reply:
Fixed in 214e0096fe954ea36918a7eaffc93eb564537210